.github/workflows/ci.yaml

```diff
@@ -18,44 +18,131 @@ jobs:
+        [[ -n "$DOCKER_USERNAME" ]] && IMAGES=${IMAGES},${IMAGE_NAME}
+        [[ -n "$QUAY_USERNAME" ]] && IMAGES=${IMAGES},quay.io/${IMAGE_NAME}
+        echo "::set-output name=images::${IMAGES}"
+
+    - name: Docker manager metadata
+      id: meta
+      uses: docker/metadata-action@v3
       with:
-        platforms: linux/amd64,linux/arm64
-        push: false
+        images: ${{ steps.image-urls.outputs.images }}
         tags: |
-          ${{env.IMAGE_NAME}}:${{github.sha}}
+          type=sha
+          type=ref,event=branch
+          type=ref,event=tag
+          type=ref,event=pr
+
+    - name: Determine if workflow should push image
+      id: should-push
+      run: |
+        SHOULD_PUSH=true
+        [[ "$GITHUB_ACTOR" == "dependabot[bot]" ]] && SHOULD_PUSH=false
+        echo "::set-output name=should-push::${SHOULD_PUSH}"
+
+    - name: Log in to ghcr.io
+      uses: docker/login-action@v1
+      if: ${{ steps.should-push.outputs.should-push == 'true' }}
+      with:
+        registry: ghcr.io
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
 
-    # the rest of these only are used if we push to master
-    - name: Login to DockerHub
-      if: github.event_name == 'push' && endsWith(github.ref,'/master')
+    - name: Log into DockerHub
       uses: docker/login-action@v1
+      env:
+        DOCKER_USERNAME: ${{ secrets.DOCKER_USERNAME }}
+      if: ${{ env.DOCKER_USERNAME != '' && steps.should-push.outputs.should-push == 'true' }}
       with:
         username: ${{ secrets.DOCKER_USERNAME }}
         password: ${{ secrets.DOCKER_PASSWORD }}
-    - name: Build and push  # theoretically, this builds what we did before, but with an unchanged dockerfile, it will just find everything in cache
-      id: docker_push
-      if: github.event_name == 'push' && endsWith(github.ref,'/master')
+
+    - name: Log into quay.io
+      uses: docker/login-action@v1
+      env:
+        QUAY_USERNAME: ${{ secrets.QUAY_USERNAME }}
+      if: ${{ env.QUAY_USERNAME != '' && steps.should-push.outputs.should-push == 'true' }}
+      with:
+        registry: quay.io
+        username: ${{ secrets.QUAY_USERNAME }}
+        password: ${{ secrets.QUAY_PASSWORD }}
+
+    - name: Build
+      id: docker_build
       uses: docker/build-push-action@v2
+      env:
+        VERSION: ${{ fromJSON(steps.meta.outputs.json).labels['org.opencontainers.image.version'] }}
       with:
+        context: .
+        push: ${{ steps.should-push.outputs.should-push == 'true' }}
+        build-args: |
+          BINARY=cloud-provider-equinix-metal
+        tags: ${{ steps.meta.outputs.tags }}
+        labels: ${{ steps.meta.outputs.labels }}
         platforms: linux/amd64,linux/arm64
-        push: true
-        tags: |
-          ${{env.IMAGE_NAME}}:${{github.sha}}
-          ${{env.IMAGE_NAME}}:master
+        cache-from: type=gha, scope=${{ github.workflow }}
+        cache-to: type=gha, mode=max, scope=${{ github.workflow }}
```

Comment:
Caching for building images. This won't save time for initial builds with new code, but saves significant time on repeated builds where non-code changes are made.